core/include/moveit/task_constructor/stages/deep_grasp_pose.h

```diff
@@ -0,0 +1,310 @@
+	 */
+	bool monitorGoal();
+
+	void activeCallback() override;
+	void feedbackCallback(const FeedbackConstPtr& feedback) override;
+	void doneCallback(const actionlib::SimpleClientGoalState& state, const ResultConstPtr& result) override;
+
+	void init(const core::RobotModelConstPtr& robot_model) override;
+	void compute() override;
+
+	void setEndEffector(const std::string& eef) { setProperty("eef", eef); }
+	void setObject(const std::string& object) { setProperty("object", object); }
+
+	void setPreGraspPose(const std::string& pregrasp) { properties().set("pregrasp", pregrasp); }
+	void setPreGraspPose(const moveit_msgs::RobotState& pregrasp) { properties().set("pregrasp", pregrasp); }
+	void setGraspPose(const std::string& grasp) { properties().set("grasp", grasp); }
+	void setGraspPose(const moveit_msgs::RobotState& grasp) { properties().set("grasp", grasp); }
+
+protected:
+	void onNewSolution(const SolutionBase& s) override;
+
+private:
+	bool found_candidates_;
+	std::vector<geometry_msgs::PoseStamped> grasp_candidates_;
+	std::vector<double> costs_;
+};
+
+template <class ActionSpec>
+DeepGraspPose<ActionSpec>::DeepGraspPose(const std::string& action_name, const std::string& stage_name,
+                                         double goal_timeout, double server_timeout)
+  : GeneratePose(stage_name), ActionBaseT(action_name, false, goal_timeout, server_timeout), found_candidates_(false) {
+	auto& p = properties();
+	p.declare<std::string>("eef", "name of end-effector");
+	p.declare<std::string>("object");
+	p.declare<double>("angle_delta", 0.1, "angular steps (rad)");
+
+	p.declare<boost::any>("pregrasp", "pregrasp posture");
+	p.declare<boost::any>("grasp", "grasp posture");
+
+	ROS_INFO_NAMED(LOGNAME, "Waiting for connection to grasp generation action server...");
+	ActionBaseT::clientPtr_->waitForServer(ros::Duration(ActionBaseT::server_timeout_));
+	ROS_INFO_NAMED(LOGNAME, "Connected to server");
+}
+
+template <class ActionSpec>
+void DeepGraspPose<ActionSpec>::composeGoal() {
+	Goal goal;
+	goal.action_name = ActionBaseT::action_name_;
+	ActionBaseT::clientPtr_->sendGoal(
+	    goal, std::bind(&DeepGraspPose<ActionSpec>::doneCallback, this, std::placeholders::_1, std::placeholders::_2),
+	    std::bind(&DeepGraspPose<ActionSpec>::activeCallback, this),
+	    std::bind(&DeepGraspPose<ActionSpec>::feedbackCallback, this, std::placeholders::_1));
+
+	ROS_INFO_NAMED(LOGNAME, "Goal sent to server: %s", ActionBaseT::action_name_.c_str());
+}
+
+template <class ActionSpec>
+bool DeepGraspPose<ActionSpec>::monitorGoal() {
+	// monitor timeout
+	bool monitor_timeout = ActionBaseT::goal_timeout_ > std::numeric_limits<double>::epsilon() ? true : false;
+	double timeout_time = ros::Time::now().toSec() + ActionBaseT::goal_timeout_;
+
+	while (ActionBaseT::nh_.ok()) {
+		ros::spinOnce();
+
+		if (found_candidates_) {
+			break;
+		}
+
+		if (ros::Time::now().toSec() > timeout_time && monitor_timeout) {
```

Comment:
Nit: I'd reverse the two conditions, so that if `monitor_timeout` is false, the other check is shortcutted entirely.

Reply:
Good catch I will move the found candidates check to an else if after the timeout